### ISSUE TITLE
Still struggling to understand errors.

### ISSUE THREAD

**Issue by tailhook** [Marwes/combine#134]
Hi, thanks for #131, it looks like an improvement. Still I have problems with how to build error handling.

For context: I usually have my own tokenizer instead of using character stream and usually tokens borrow chunks from source string (`Token<'a> { value: &'a str }`)

1. I find error API too complex. First there are setters, then there are getters and there are more convertions. I don't understand how parser will use them.
2. I don't understand when and how error converts from `StreamError` to `ParseError` 
3. Finally I want to expose a custom error from my module, without exposing `Token` type. The best way I've found is to use just to use Error and Errors from easy module and just format them to string. It looks like suboptimal though.
4. **update:** forgotten another API issue: there is a constructor for end_of_input, but not a way to check. Is `== Error::end_of_input()` a good idea?

Are there any good examples of error handling?

**Comment by Marwes** [comment]
> I find error API too complex. First there are setters, then there are getters and there are more convertions. I don't understand how parser will use them.

First of all, the current API is more or less based on the `easy` module https://docs.rs/combine/3.0.0-beta.1/combine/easy/index.html. If you just need good error reporting and don't need to supply your own errors then just using the error defined there is a good idea.

There could certainly be some improvements to it however so I will look into to that.

> I don't understand when and how error converts from StreamError to ParseError

[StreamError](https://docs.rs/combine/3.0.0-beta.1/combine/error/trait.StreamError.html) represents a single error that has happened during parsing, whether that be something *unexpected* in the stream or a report that something else was *expected* or just some generic error message.

> Finally I want to expose a custom error from my module, without exposing Token type. The best way I've found is to use just to use Error and Errors from easy module and just format them to string. It looks like suboptimal though.

I think what you want is the [map_token](https://docs.rs/combine/3.0.0-beta.1/combine/stream/easy/struct.Errors.html#method.map_token) method which lets you transform the token type of the error into something else (there are also `map_position` and `map_range`

> update: forgotten another API issue: there is a constructor for end_of_input, but not a way to check. Is == Error::end_of_input() a good idea?

That is how I do it, though perhaps it would be a good idea to add a default as `is_end_of_input` as well.

> Are there any good examples of error handling?

I should probably write one. So far I haven't had any need for anything more than the `easy` module (and you are the first to ask about it). The whole error rewrite has thus been mainly about making the `easy` module optional so that combine can be used with `no_std`.

**Comment by tailhook** [comment]
> StreamError represents a single error that has happened during parsing, whether that be something unexpected in the stream or a report that something else was expected or just some generic error message.

Yes that look kinda obvious. Still `parse` returns Stream error and `easy_parse` doesn't work on custom streams. Am I missing something?

Also if `parse` returns just one error, how would I get `unexpected '{' , expected ',', ':' ` style rror? Maybe I should use something other than `parse`  for top-level parsing function?

>>  update: forgotten another API issue: there is a constructor for end_of_input, but not a way to check. Is == Error::end_of_input() a good idea?
>
> That is how I do it, though perhaps it would be a good idea to add a default as is_end_of_input as well.

The confusing thing is the code in `easy::Error`:
```
    pub fn end_of_input() -> Error<T, R> {
        Error::Unexpected("end of input".into())
    }
```
It looks like allocating a string and comparing the whole string. Even if it's optimized by the compiler, still doesn't look good.

> I think what you want is the map_token method which lets you transform the token type of the error into something else (there are also map_position and map_range

Yes, that's helpful. Thanks!

> So far I haven't had any need for anything more than the easy module (and you are the first to ask about it). The whole error rewrite has thus been mainly about making the easy module optional so that combine can be used with no_std.

Well, that's okay. Probably just should be documented in the docs. Like, if you don't need `no_std`just use easy error. Also in my mind `easy` module is clearly associated with parsing characters from `easy::Stream` so it took time for me to find out that I can use errors from that. May be it's better to move that error implementation into other module. E.g. `error` itself?

And I find `Info`  (and that it's different in `easy` and `error`) confusing. I think this structure can be hidden entirely behind constructors of StreamError

**Comment by Marwes** [comment]
> Yes that look kinda obvious. Still parse returns Stream error and easy_parse doesn't work on custom streams. Am I missing something?

[parse](https://docs.rs/combine/3.0.0-beta.1/combine/trait.Parser.html#method.parse) returns a full `ParseError` actually. I guess it might be a bit confusing that `StreamOnce::Error` is actually a full `ParseError` while it is only `uncons` that return the smaller `StreamError`.

[easy_parse](https://docs.rs/combine/3.0.0-beta.1/combine/trait.Parser.html#method.easy_parse) actually sort of works on any stream. As long as you define your functions to be generic over the stream type then calling `easy_parse` on such a generic function (`I: Stream`) "just works" when you give it your actual stream since in that case all it does is wrap that stream internally. (I will try to think of where I can explain this more clearly.

You can also always get the same effect as `easy_parse` by just wrapping your stream with `easy::Stream` as well and call `.parse`.

> It looks like allocating a string and comparing the whole string. Even if it's optimized by the compiler, still doesn't look good.

It actually stores the `&'static str` as-is. The comparison is still slow I will admit but since you can only reach the end of input once during a single parse you will at worst do that comparison a handful of times (if it is encountered during a `choice` parser). So it shouldn't affect performance in any noticeable way.

Still, it is an argument in favor of an `is_end_of_input` method since that would only need to check the pointer.

> Also in my mind easy module is clearly associated with parsing characters from easy::Stream so it took time for me to find out that I can use errors from that.

I think will add an example onhow to use a custom tokenizer, that should probably clarify things.

> Maybe it's better to move that error implementation into other module. E.g. error itself?

Yeah, maybe I was a bit torn on the placement. On one hand it is nice to keep the whole implementation in a single module. On the other hand, the only thing it does is modify which error are returned. But how you actually use `easy` is as a `Stream` so it also makes sense to keep it under `stream`... Perhaps it should only be a standalone module, separate from both `stream` and `error`

> And I find Info (and that it's different in easy and error) confusing. I think this structure can be hidden entirely behind constructors of StreamError

Not sure how to resolve this really. I don't want to hide anything about the `easy::Error` structure since I want users to be able to traverse to transform it to a different error type or to just print it in a different format.

**Comment by tailhook** [comment]
Okay. Also it might happen that some frustrations are because I've copied pieces of code from previous project (on old combine) :)

> parse returns a full ParseError actually. I guess it might be a bit confusing that StreamOnce::Error is actually a full ParseError while it is only uncons that return the smaller StreamError.

Yes, that was very counter-intuitive. Maybe it's better to turn associated types other way around?

Another part of confusion is that `easy::Error` implements both traits.

Overall, thanks for explanations! I figured out most of the error reporting I want so far.

**Comment by Marwes** [comment]
> Okay. Also it might happen that some frustrations are because I've copied pieces of code from previous project (on old combine) :)

If there is something non-obvious to fix which is not covered by https://github.com/Marwes/combine#breaking-changes please let me know so I can add it!

> Yes, that was very counter-intuitive. Maybe it's better to turn associated types other way around?

That seems backward in a different way though since `ParseError`s contain `StreamError`s which is reflected in how the traits are set up now :/. I may try to find `StreamError` perhaps that would help.

> Another part of confusion is that easy::Error implements both traits.

Yeah, that isn't strictly necessary but it mirrors how the `no_std` errors for `&[T]` and `&str` are set up so it seems inconsistent not to have it. I'd like to have just a blanket `impl StreamError for T where T: ParseError` but coherence does not allow it.

**Comment by tailhook** [comment]
By the way, [here is the project](https://crates.io/crates/graphql-parser) I've been doing.

I'm not sure anything actionable left here, except doc/example.

Any plans for release? Since depending on beta is somewhat ugly.

**Comment by Marwes** [comment]
Looked through the repo a bit to see if there was things I could improve on combine's side. Noticed a few things you may want to look at tweaking.

* I'd lock down the version in `Cargo.toml` to `=3.0.0-beta.1`. Apparently `3.0.0-beta.2`, `3.0.0-gamma.123` are all semver compatible which is a bit hazardous when it comes to pre-releases (which I just learnt 2 days ago). Would be more useful if all pre-releases were considered semver incompatible unless explicitly opted in to.

* https://github.com/tailhook/graphql-parser/blob/088eb936700b2b13a746a5b580400a6ae2b2ecbd/src/query_grammar.rs#L221-L228
You could use [choice](https://docs.rs/combine/3.0.0-beta.1/combine/fn.choice.html) here (and a couple of other places). It can take a tuple of parsers as an alternative to chaining `or`.

* https://github.com/tailhook/graphql-parser/blob/088eb936700b2b13a746a5b580400a6ae2b2ecbd/src/query_grammar.rs#L252-L254

* Any reason why you refrained from using [parser!](https://docs.rs/combine/3.0.0-beta.1/combine/macro.parser.html)? There is nothing wrong with using freestanding functions to define parsers as you do, but the macro automatically does the `parse_lazy+add_error` optimization (which I saw that you did in the [helpers module](https://github.com/tailhook/graphql-parser/blob/088eb936700b2b13a746a5b580400a6ae2b2ecbd/src/helpers.rs#L79-L118). (This will be a moot point in half a year or so once `impl Trait` gets stabilized and I will be very happy to deprecate the macro).

> Any plans for release? Since depending on beta is somewhat ugly.

Not quite yet, I need to make sure that #116 can be added in a backwards compatible manner (or implement it before release), I need to make a decision on #136 and I will probably do some changes to the error traits (as prompted in this issue). Hopefully that means I can do a release in 1-2 months.

**Comment by tailhook** [comment]
> all semver compatible which is a bit hazardous when it comes to pre-releases 

Didn't know. Will fix it.

> You could use choice here (and a couple of other places). It can take a tuple of parsers as an alternative to chaining or.

Well, I'm not sure why I did not use choice. Probably because I've discovered `or` earlier and stuck with it. I've checked that `choice` gives no performance benefits so.... Will try on new code :)

> Any reason why you refrained from using parser!?

I didn't see it before. But generally I avoid macros if possible. Looking at it now, it's super-weird. I understand that's because macro syntax is limited. So if we have `impl Trait` soon, there so reason to use it. (Additionally, these helpers are with me probably since combine <1.0, so it was just easier to copy them)
